massa-bootstrap/src/server.rs

```
@@ -0,0 +1,455 @@
+        let cache_timeout = self.bootstrap_settings.cache_duration.to_duration();
+        let mut bootstrap_data: Option<(
+            ExportProofOfStake,
+            BootstrapableGraph,
+            BootstrapPeers,
+            Arc<RwLock<FinalState>>,
+        )> = None;
+        let cache_timer = sleep(cache_timeout);
+        let per_ip_min_interval = self.bootstrap_settings.per_ip_min_interval.to_duration();
+        tokio::pin!(cache_timer);
+        /*
+            select! without the "biased" modifier will randomly select the 1st branch to check,
+            then will check the next ones in the order they are written.
+            We choose this order:
+                * manager commands to avoid waiting too long to stop in case of contention
+                * cache timeout to avoid skipping timeouts cleanup tasks (they are relatively rare)
+                * bootstrap sessions (rare)
+                * listener: most frequent => last
+        */
+        loop {
+            massa_trace!("bootstrap.lib.run.select", {});
+            tokio::select! {
+                // managed commands
+                _ = self.manager_rx.recv() => {
+                    massa_trace!("bootstrap.lib.run.select.manager", {});
+                    break
+                },
+
+                // cache cleanup timeout
+                _ = &mut cache_timer, if bootstrap_data.is_some() => {
+                    massa_trace!("bootstrap.lib.run.cache_unload", {});
+                    bootstrap_data = None;
+                }
+
+                // bootstrap session finished
+                Some(_) = bootstrap_sessions.next() => {
+                    massa_trace!("bootstrap.session.finished", {"active_count": bootstrap_sessions.len()});
+                }
+
+                // listener
+                Ok((dplx, remote_addr)) = listener.accept() => if bootstrap_sessions.len() < self.bootstrap_settings.max_simultaneous_bootstraps as usize {
```

`as usize` conversions should be avoided, there is a risk of losing bits which is a silent error. `from` or `try_into()` even with a following `expect()` are better IMHO because we know were it comes from if it ever fails. https://doc.rust-lang.org/rust-by-example/types/cast.html

---

Done.